### PR TITLE
feat(Core/AI): CU_SAI - Custom Action Options + Polar Coords System Offset Relocating

### PR DESCRIPTION
**Pull request by P-Kito**
* SMART_ACTION_CUSTOM_CAST
* SMART_ACTION_CONE_SUMMON (Legion, BFA, Mop like Custom Cone-Spells Yay!)
* SMART_ACTION_PLAYER_TALK (Custom DBM "Bomb Bla on me in 3...")
* SMART_ACTION_VORTEX_SUMMON (Vortex = Spiral)
* SMART_ACTION_CU_ENCOUNTER_START (Removes cds on boss start, retail like!)
* **MODIFIED:** SMART_ACTION_SET_ORIENTATION (Allow a random orientation)

```
    SMART_ACTION_CUSTOM_CAST                        = 218,    // spellId, castflag, bp0, bp1, bp2
    SMART_ACTION_CONE_SUMMON                        = 219,    // entry, duration (0 = perm), dist between rings, dist between earch summon in a row, length of cone, width of cone (angle)
    SMART_ACTION_PLAYER_TALK                        = 220,    // creature_text.id, yell? (0/1)
    SMART_ACTION_VORTEX_SUMMON                      = 221,    // entry, duration (0 = perm), spiral scaling, spiral appearance, range max, phi_delta     <-- yes confusing math, try it ingame and see, my lovely AC boys!
    SMART_ACTION_CU_ENCOUNTER_START                 = 222,    // Resets cooldowns on all targets and removes Heroism debuff(s)
```

## TESTS PERFORMED:
- PLEASE RETEST, EXPERIMENT WITH THOSE SPIRAL/VORTEX, CONE SUMMONS! ;-)

## HOW TO TEST THE CHANGES:
- Create a SAI for a creature and use those new actions... Please ask your question. A "Vortex" is a "Spiral" - Just play with those actions honestly, you will understand.

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: src/server/game/AI/SmartScripts/SmartScriptMgr.h [hunk 1]
```diff
@@ -590,7 +590,13 @@ enum SMART_ACTION
     SMART_ACTION_MUSIC                              = 216,    // SoundId, onlySelf, type
     SMART_ACTION_RANDOM_MUSIC                       = 217,    // SoundId1, SoundId2, SoundId3, SoundId4, onlySelf, type
 
-    SMART_ACTION_AC_END                             = 218,    // placeholder
+    SMART_ACTION_CUSTOM_CAST                        = 218,    // spellId, castflag, bp0, bp1, bp2
+    SMART_ACTION_CONE_SUMMON                        = 219,    // entry, duration (0 = perm), dist between rings, dist between earch summon in a row, length of cone, width of cone (angle)
+    SMART_ACTION_PLAYER_TALK                        = 220,    // creature_text.id, yell? (0/1)
```
Comment (P-Kito) [inline src/server/game/AI/SmartScripts/SmartScriptMgr.h:595]: This is not right yet, its  trinity_strings  DB ... I originally wanted to do it with creature_text

**Comment by P-Kito** [comment]
I will test it on my own in the next days... I just need to setup database and stuff

**Comment by P-Kito** [comment]
[x] Custom cast tested ingame
[x] Player talk tested ingame
[x] Cone summon tested ingame
grafik
[x] Spiral/Vortex Summon tested ingame 
Values: 
221 - action
38163 - creature id 
5000 - summon duration 
5 - Spiral Scaling 
25 - spiral "appearance 
60 - range max 
40 - phi delta (check math on ur own... spiral parameters)

File: src/server/game/Entities/Object/Object.cpp [hunk 2]
```diff
@@ -881,13 +881,22 @@ void Object::RemoveByteFlag(uint16 index, uint8 offset, uint8 oldFlag)
...
 
+void Position::RelocatePolarOffset(float angle, float dist, float z /*= 0.0f*/)
+{
+    SetOrientation(GetOrientation() + angle);
+
+    m_positionX = GetPositionX() + dist * std::cos(GetOrientation());
+    m_positionY = GetPositionY() + dist * std::sin(GetOrientation());
+    m_positionZ = GetPositionZ() + z;
+}
+
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Object/Object.cpp:899]: cool stuff!

**Comment by FrancescoBorzi** [comment]
nice work @P-Kito , can you please copy paste this nice explanation somewhere into the [wiki](https://github.com/azerothcore/wiki) ?  otherwise it will get lost in github :)

**Comment by Si1ker** [comment]
This is gonna be my favorite PR so far, nice work @P-Kito !! ❤️

**Comment by BarbzYHOOL** [comment]
i'm super happy that you finally let you tempted to the bright side of the force Kito! Looks awesome

Note: i had few ideas for SAI, maybe u'd be interested in adding them too in another PR? :D

**Comment by P-Kito** [comment]
Send me your ideas on Discord, I will see ;-)

**Pull request merged**